### DISCUSSION TITLE
The future twinBASIC IDE

### DISCUSSION THREAD

**Discussion by WaynePhillipsEA** [twinbasic/twinbasic#763] [General]
This post is to outline the new direction I'm setting out for the future twinBASIC IDE.

As you know, my long term plan is to move the IDE away from VS Code to Theia (#458).  This will give us much more control over the IDE UX and the plumbing that goes on between the compiler and the IDE.  Theia offers great potential for the future of twinBASIC, and I still expect this move to happen in the distant future, but the reality is that Theia is not ready for us yet as it is lacking several key features that we use from VS Code, and some of the vscode API that is included isn't mature/stable enough for us at this moment in time.

As mentioned over on Discord, over the last few days I've had a play around with Webview2 in twinBASIC, and for my experiment I mocked up a very simple IDE design, somewhat based on VS Code, and most importantly using the open-source Monaco editor, which is the exact same text editor as used by VS Code.  I was pleasantly surprised at how quickly I was able to put something together that *looked* somewhat like an IDE.  With using Monaco as the editor, it will provide an easy route to implement all of the compiler features that we are accustomed to.

The mockup is not a functioning IDE yet... but I expect I could get most parts of the IDE hooked up to the compiler services within a couple of weeks.  The are several key advantages of going this route;

- The new IDE will be very much simpler than VS Code, designed specifically for our needs
- Startup will be much quicker and more stable than in VS Code due to being a first-class citizen rather than an extension
- There wont be any battling with the workspace-trust feature of VS Code
- We are free to create custom top-level menus and toolbars, just like in VBx
- We are free to create modal dialogs when needed, e.g. mimicking the VBx New/Existing startup dialog
- We can remove the need for the additional code-workspace file needed for each project
- Being a pure HTML/js solution, we could allow for local debugging even on older systems such as Windows XP/2K via their webbrowser (localhost)
- Cross platform will also be easily supported as we will be using 100% web-based tech for the IDE side of things
- It offers a good opportunity to refactor some of the javascript code that binds the compiler to the IDE, ready for the eventual move to Theia

I hope you agree with my decision that it worth spending a few weeks on this very soon, as I suspect many more developers will look at twinBASIC once we can offer an IDE that is more on par with VBx in terms of being simple and intuitive to use.

<hr/>

To summarize, the plan is to now create a new, simple IDE built on cross-platform web-based technologies for our immediate needs.  The VS Code extension will continue to be maintained and improved for those that want to use it that way.  The long-term plan is to move to Theia, though we will probably keep the new 'simple' IDE available as it has significantly fewer system requirements.

**Comment by WaynePhillipsEA** [comment]
![tbNewIDEMockup1](https://user-images.githubusercontent.com/25738109/155891274-51fde52d-4928-4e70-8652-cb57be14756a.gif)

Here's the mockup in action, just showing the simplified layout, and the start of work on things like the tree lists, all working from a Webview2 control, in a twinBASIC form.

**Comment by mwolfe02** [comment]
How do the .twinproj files factor in to the new IDE?  Will the virtual filesystem still be a requirement?  At one point you had mentioned that eventually .twinproj files could be more like solution files that didn't have embedded source code (i.e., all the source code would be in external files outside of the .twinproj containers).  I guess I'm wondering whether the virtual filesystem was needed to play nice with VSCode or if it's more of a convenience for the twinBASIC compiler?

**Reply by WaynePhillipsEA** [reply]
I probably won't change this to start with, but it's still planned to offer external file support (with the twin proj file then acting much like an ordinary dumb project file, like .VBP in VBx).

**Reply by DinyaZ** [reply]
Very, very good news!! A custom tailored, simple and user friendly IDE was badly needed. Perhaps VS Code was okey for testing and developing, but seemed to be very overcomplicated for real life work. 
Thank you, Wayne!

**Comment by FullValueRider** [comment]
And, of course, the 64 million dollar question is..... could this be done in such a way that the editor could be hosted by office apps thus replacing the built in Ide.

**Reply by WaynePhillipsEA** [reply]
Absolutely!  That's the plan!

**Reply by mburns08109** [reply]
bootstrap the tBIDE in VBA?

**Reply by mansellan** [reply]
actually, I'd argue that this _is_ a case of dogfooding. Bootstrapping is specifically about building a compiler in its own language IIUC.

**Comment by dmrvb** [comment]
This is probably good, I don't fully understand it, but it might be a distraction?
VS code works and has some nice features, but a few major irritations. TwinBasic is not yet  a VB replacement because it still lacks many GUI features, though it looks to be getting tantalisingly close. I would suggest a bit more work on the controls such that a VB project with a reasonably complex GUI can be imported and compiled. At this point TwinBasic really is a "VB replacement" even with some rough edges and limitations. At that stage you can maybe look at the IDE and making it more VB friendly?
I confess I have a vested interest because I really want to get a standalone GUI app compiled in TwinBasic.

**Comment by camomille22003** [comment]
A great hope to find our favorite IDE VB6 but with all possibilities of modernization, here is the result for the Arduino IDE, it's a treat ,exist for Windows 64,Linux 64,MacOs 64
**" _The Arduino IDE 2.x is a major rewrite, sharing no code with the IDE 1.x. It is based on the Theia IDE framework and built with Electron.  
This new IDE was developed with the goal of preserving the same interface and user experience of the previous major version in order to provide a frictionless upgrade._"**

![Image4](https://user-images.githubusercontent.com/33807881/156318803-674b3f4f-caae-4714-82cd-6c8cb85ab6c1.jpg)

**Comment by WaynePhillipsEA** [comment]
A first proper look at the work-in-progress twinBASIC IDE...  debugger is fully integrated now.  Code editor has a lot of missing features which will be worked on this coming week.
![tbNewIde1](https://user-images.githubusercontent.com/25738109/156914959-6e0d97e5-46e8-4567-b3ac-b9fd7314dcc9.gif)

**Reply by wqweto** [reply]
Wow, this looks like a lot of unplanned work! Feature creep is real! :))

Hope this turns out to be major “selling point” for the product adoption.

**Comment by WaynePhillipsEA** [comment]
Another sneak peek....  now supporting unsaved projects (like VB6), and most of the compiler services are now functional (semantic highlighting, codelens, intellisense, etc)

![tbNewIde2](https://user-images.githubusercontent.com/25738109/158009725-c18e583b-321f-4769-9dd9-d585a8698066.gif)

**Reply by DinyaZ** [reply]
Wayne, thank you very much for this promising new development!
.
I know that this is in the very early phase now, but take care of the design also, and try to have a unified, modern UI in the whole project. Same color scheme, unified rectangular buttons, etc. For example please give "dark theme" to the form designer, instead of the shading effect and oval "Preview" button.

**Reply by WaynePhillipsEA** [reply]
Thanks. It is planned to work on the form designer UX at some point, but it won't be changed for the first release.

**Comment by WaynePhillipsEA** [comment]
We're getting very close now ...

![tbNewIdeDemo3](https://user-images.githubusercontent.com/25738109/160071586-d43204c5-0f8c-4ffd-8574-f5018820eb79.gif)

**Reply by wqweto** [reply]
Btw, which panes are HTML and which are native?

**Reply by WaynePhillipsEA** [reply]
They're all HTML

**Reply by wqweto** [reply]
Nice! So this should be as portable as an Electron based desktop app?

**Reply by WaynePhillipsEA** [reply]
Yup, it will be easily portable.  At the moment it's hosted in a WebView2 container on a tB form, but changing to Electron (or even just direct access via any webrowser) would be easy.

**Reply by mburns08109** [reply]
OK, in for a penny, in for a pound on this question...will you consider sharing the tB source for this IDE? (Which, I hope you wrote in tB, as I thought I heard you say you intended to do...?)

**Reply by WaynePhillipsEA** [reply]
As mentioned above, it's a HTML-app, hosted within a tB form (using a WebView2 control).  The tB host side of things mainly just exposes some OS functionality (file dialogs, and registry storage) to the HTML/javascript app.    But ultimately yes, the whole thing will be put on Github once it's more complete and has been tidied up.

**Reply by mansellan** [reply]
Can I just check -  the new IDE is still using LSP and DAP to talk to the back-end right?

**Reply by WaynePhillipsEA** [reply]
That's a slightly complicated one to answer so bear with me.  Technically yes, but we are using some custom requests (which are allowed per the specs) to _bundle_ requests for optimization purposes, because our new IDE knows that the backend compiler services can handle them together without extra cost.  Bundling the requests means much less 'chatter' between the server and client, and this provides a smoother overall user experience with less chance of the server/client becoming out of sync.   

The compiler of course still supports the unbundled requests for VS code support, but the new IDE _currently_ only supports the bundled custom requests, and so it is more tightly tied to our compiler... for now at least.

**Comment by bclothier** [comment]
> registry storage

Just to clarify --- should the IDE be even using the Windows registry at all? The trend has been to move away from the registry for anything, using alternatives (e.g. config or manifest files) and as a model, the IDE shouldn't need to use Windows registry at all except for working with COM?

**Reply by WaynePhillipsEA** [reply]
It's currently used for IDE settings.  I'll be moving it over to AppData storage in future, but for now I took the easy option :)

**Comment by WaynePhillipsEA** [comment]
The first beta of the IDE has now been released.  Further discussion on this topic should continue over on #772.
